pkg/mod/mod.go

```diff
@@ -0,0 +1,67 @@
+package mod
+
+import (
+	"runtime/debug"
+	"sync"
+
+	"golang.org/x/mod/module"
+	"golang.org/x/mod/semver"
+)
+
+var (
+	buildInfoOnce = sync.Once{}
+	buildInfo     *debug.BuildInfo
+)
+
+func GetVersion(name string) (modVersion string) {
+	return moduleVersion(name, readBuildInfo())
+}
+
+func moduleVersion(name string, bi *debug.BuildInfo) (modVersion string) {
+	if bi == nil {
+		return
+	}
+	// iterate over all dependencies and find buildkit
+	for _, dep := range bi.Deps {
+		if dep.Path != name {
+			continue
+		}
+		// get the version of buildkit dependency
+		modVersion = dep.Version
+		if dep.Replace != nil {
+			// if the version is replaced, get the replaced version
+			modVersion = dep.Replace.Version
+		}
+		if !module.IsPseudoVersion(modVersion) {
+			return
+		}
+		// if the version is a pseudo version, get the base version
+		// e.g. v0.10.7-0.20230306143919-70f2ad56d3e5 => v0.10.6
+		if base, err := module.PseudoVersionBase(modVersion); err == nil && base != "" {
+			// set canonical version of the base version (removes +incompatible suffix)
+			// e.g. v2.1.2+incompatible => v2.1.2
+			base = semver.Canonical(base)
+			// if the version is a pseudo version, get the revision
+			// e.g. v0.10.7-0.20230306143919-70f2ad56d3e5 => 70f2ad5
+			if rev, err := module.PseudoVersionRev(modVersion); err == nil && rev != "" {
+				// append the revision to the version
+				// we set the short commit sha instead of the uncommon
+				// one (https://github.com/golang/go/issues/34745)
+				// e.g. v0.10.7-0.20230306143919-70f2ad56d3e5 => v0.10.6+70f2ad5
+				modVersion = base + "+" + rev[:7]
```

Comment:
Without actually consulting the Git repository, there's not enough information to know whether seven characters is long enough to unambiguously identify a single commit. I have personally been burned by a situation where a repo grew large enough that seven characters was insufficient. Please do not truncate the commit hash further than Go modules already do.